src/TesApi.Web/Events/RunnerEventsProcessor.cs

```diff
@@ -0,0 +1,516 @@
+                    Validate(false, $"{nameof(content.Name)}('{content.Name}') is not recognized.");
+                    break;
+            }
+
+            return message;
+
+            void ValidateFailedStatus()
+            {
+                if (Tes.Runner.Events.EventsPublisher.FailedStatus.Equals(content.StatusMessage))
+                {
+                    Validate(content.EventData.ContainsKey("errorMessage"), $"{nameof(content.Name)}('{content.Name}' with {nameof(Tes.Runner.Events.EventsPublisher.FailedStatus)}) does not contain 'errorMessage'");
+                }
+            }
+
+            void ValidateCreated()
+                => Validate(content.Created != default, $"{nameof(content.Name)}('{content.Name}') {nameof(content.Created)} was not set.");
+
+            static void Validate([System.Diagnostics.CodeAnalysis.DoesNotReturnIf(false)] bool condition, string message)
+            {
+                if (!condition)
+                {
+                    throw new DownloadOrParseException(message);
+                }
+            }
+        }
+
+        private static readonly IReadOnlyDictionary<string, int> EventsInOrder = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
+        {
+            { Tes.Runner.Events.EventsPublisher.DownloadStartEvent, int.MinValue },
+            { Tes.Runner.Events.EventsPublisher.DownloadEndEvent, int.MinValue + 1 },
+            { Tes.Runner.Events.EventsPublisher.ExecutorStartEvent, -1 },
+            { Tes.Runner.Events.EventsPublisher.ExecutorEndEvent, +1 },
+            { Tes.Runner.Events.EventsPublisher.UploadStartEvent, int.MaxValue - 1 },
+            { Tes.Runner.Events.EventsPublisher.UploadEndEvent, int.MaxValue },
+        }.AsReadOnly();
+
+        /// <summary>
+        /// Returns a sequence in the order the events were produced.
+        /// </summary>
+        /// <typeparam name="T"><paramref name="source"/>'s enumerated type.</typeparam>
+        /// <param name="source">Unordered enumeration of events.</param>
+        /// <param name="messageGetter">Function that returns <see cref="RunnerEventsMessage"/> from <typeparamref name="T"/>.</param>
+        /// <returns>Ordered enumeration of events.</returns>
+        /// <remarks>This method assumes every <see cref="RunnerEventsMessage"/> was successfully validated by <see cref="ValidateMessageMetadata(RunnerEventsMessage)"/>.</remarks>
+        public IEnumerable<T> OrderProcessedByExecutorSequence<T>(IEnumerable<T> source, Func<T, RunnerEventsMessage> messageGetter)
+        {
+            ArgumentNullException.ThrowIfNull(source);
+            ArgumentNullException.ThrowIfNull(messageGetter);
+
+            return source.OrderBy(t => OrderBy(messageGetter(t))).ThenBy(t => ThenBy(messageGetter(t)));
+
+            static DateTime OrderBy(RunnerEventsMessage message)
+                => (message.RunnerEventMessage?.Created ?? DateTime.Parse(message.Tags["created"])).ToUniversalTime();
```

Comment:
The created property should never be null, so we should declare the type as such.

Reply:
Defense in depth. Doesn't cost much to never run code that won't be hit.